libsql-sqlite3/src/vectordiskann.c

```diff
@@ -345,4 +403,1048 @@ void nodeBinDebug(const DiskAnnIndex *pIndex, const BlobSpot *pBlobSpot) {
+    assert(pCtx->aDistances[i] <= pCtx->aDistances[i + 1]);
+  }
+#endif
+  for(i = 0; i < pCtx->nCandidates; i++){
+    DiskAnnNode *pCandidate = pCtx->aCandidates[i];
+    if( pCandidate->visited ){
+      continue;
+    }
+    return i;
+  }
+  return -1;
+}
+
+/**************************************************************************
+** DiskANN core
+**************************************************************************/
+
+// return position to insert new edge(C) for a node or -1 if we should ignore it
+// we also check that no current edge(B) will "prune" new vertex: i.e. dist(B, C) >= (means worse than) alpha * dist(node, C) for all current edges
+// if any edge(B) will "prune" new edge(C) we will ignore it (return -1)
+static int diskAnnInsertEdgeIdx(const DiskAnnIndex *pIndex, BlobSpot *pNodeBlob, const Vector *pNewVector) {
+  int i, nEdges, nMaxEdges, iInsert = -1;
+  Vector nodeVector, edgeVector;
+  float nodeToNew;
+
+  nEdges = nodeBinEdges(pIndex, pNodeBlob);
+  nMaxEdges = nodeEdgesMaxCount(pIndex);
+  nodeBinVector(pIndex, pNodeBlob, &nodeVector);
+  nodeToNew = diskAnnVectorDistance(pIndex, &nodeVector, pNewVector);
+
+  for(i = nEdges - 1; i >= 0; i--){
+    float edgeToNew, nodeToEdge;
+    
+    nodeBinEdge(pIndex, pNodeBlob, i, NULL, &edgeVector);
+    edgeToNew = diskAnnVectorDistance(pIndex, &edgeVector, pNewVector);
+    nodeToEdge = diskAnnVectorDistance(pIndex, &nodeVector, &edgeVector);
+    if( nodeToNew > pIndex->pruningAlpha * edgeToNew ){
+      return -1;
+    }
+    if( nodeToNew < nodeToEdge ){
+      iInsert = i;
+    }
+  }
+  if( iInsert == -1 && nEdges < nMaxEdges ){
+    return nEdges;
+  }
+  return iInsert;
+}
+
+// prune edges after we inserted new edge at position iInserted
+// we only need to check for edges which will be pruned by new vertex
+// no need to check for other pairs as we checked them on previous insertions
+static void diskAnnPruneEdges(const DiskAnnIndex *pIndex, BlobSpot *pNodeBlob, int iInserted) {
+  int i, s, nEdges;
+  Vector nodeVector, hintVector;
+
+  nodeBinVector(pIndex, pNodeBlob, &nodeVector);
+  nEdges = nodeBinEdges(pIndex, pNodeBlob);
+
+  assert( 0 <= iInserted && iInserted < nEdges );
+
+#if defined(SQLITE_DEBUG) && defined(SQLITE_VECTOR_TRACE)
+  DiskAnnTrace(("before pruning:\n"));
+  nodeBinDebug(pIndex, pNodeBlob);
+#endif
+
+  nodeBinEdge(pIndex, pNodeBlob, iInserted, NULL, &hintVector);
+
+  // remove edges which is no longer interesting due to the addition of iInserted
+  i = 0;
+  while( i < nEdges ){
+    Vector edgeVector;
+    float nodeToEdge, hintToEdge;
+    if( i == iInserted ){
+      i++;
+      continue;
+    }
+    nodeBinEdge(pIndex, pNodeBlob, i, NULL, &edgeVector);
+    nodeToEdge = diskAnnVectorDistance(pIndex, &nodeVector, &edgeVector);
+    hintToEdge = diskAnnVectorDistance(pIndex, &hintVector, &edgeVector);
+    if( nodeToEdge > pIndex->pruningAlpha * hintToEdge ){
+      nodeBinDeleteEdge(pIndex, pNodeBlob, i);
```

Comment:
This is quadratic, no? It feels that we could delete edges and compress them only once to not move the elements too many time, no?

Reply:
Yes, fixed this by changing `nodeBin{Delete,Insert}Edge` implementations: now insert just replace edge at the position, and delete just swap current edge with last edge in the array